### PR TITLE
emit resumeConnection event from container

### PR DESCRIPTION
**Pull request by markfields**
Our internal Loop partners are trying to measure performance and reliability when establishing the websocket connection.  However, there is no centralized place to detect when we are actually starting to connect.

It could be from them calling `resume`, or `setAutoreconnect`, or some time after "disconnect" event is raised as the framework tries to reconnect.  And in the case of `resume`, one supported configuration is to let the host app call `resume`, in which case the shared code being instrumented has _no_ visibility into when the connection attempt begins.

So, this PR adds a new "resumeConnection" event which is fired from the Container when it is kicking off the connection.

My intention would be to add this event to `IContainEvents` in `container-definitions` package once this implementation goes in.

### CODE REVIEW

**Comment by markfields** [comment]
@vladsud I need you to look at this please.  Thanks!

**Comment by vladsud** [comment]
I'm not sure what hosts are looking for exactly, but would same problem be solved by exposing duration on existing events? I.e. what we log in telemetry as _duration_ in _Connecting & _Connected events?

Note that either way time covered will include things like data center meltdown (in various forms, including 429s), so it will never cover single attempt. So hosts are likely to want to cover user experiences only, and even today they have full control here (in the form of where they call various APIs) - I think it would be easier for hosts to use explicit workflows and track individual scenarios (connection on boot, connection on user returning back to computer), then rely on general events like that.

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -1036,6 +1044,9 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             args.mode = "write";
         }
 
+        // Notify listeners that we are attempting to resume the delta connection
+        this.emit("resumeConnection", args);
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1048]: I do not think there is guarantee that we get here when disconnected. For example, Container.resume() can be called any number of times in a row.

Comment (markfields) [reply packages/loader/container-loader/src/container.ts:1048]: Hmmm great point. I think just checking connection state somehow for disconnected should be ok.

If someone calls resume twice in a row, and the first one hasn't yielded a connection yet, I think it's fair to emit this event twice. The caller would be listening for the connected event, and would be able to easily dedupe them. But I agree that once connected, another call to resume should not result in this event.

Comment (markfields) [reply packages/loader/container-loader/src/container.ts]: Updated it - and this should probably move to be managed by connectionStateHandler, maybe along with #7884

**Comment by markfields** [comment]
> I'm not sure what hosts are looking for exactly, but would same problem be solved by exposing duration on existing events?

@vladsud - This is an ask coming out of the work I've done with Vivek, we are trying to instrument initial connection in OWH code.  Current reality is that many sessions close without ever connecting.  So we need a signal when we are trying to start to get the baseline of our reliability goal.

**Comment by markfields** [comment]
>  even today they have full control here (in the form of where they call various APIs)

See my initial comment.  There is a case where host apps can take responsibility for calling `resume` and OWH doesn't know.  This is what triggered this work.  Hoping to get this in for next release (should have started it sooner) so it's ready for Vivek to code against when he returns in 2 weeks.

**Comment by markfields** [comment]
Other way to think about this - The host is interested in cases where the user is disconnected but doesn't want to be.  Today, it's starting a 15 second timer after `disconnected` event.  But that's not "fair" to us, because maybe we don't even try to connect until some time later.  This event gives the host a clear picture - "I am trying to connect....I connected" will be the message they get now.

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -1036,6 +1044,13 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             args.mode = "write";
         }
 
+        if (this.connectionState === ConnectionState.Disconnected) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1047]: I do not think this check is correct one. Client might be disconnected but trying to connect. I.e., if you call this API 10 times in a row in a tight loop, 10 events will be fired. I do not think that's what clients want.

The problem with events like that is that you will likely need to expose (eventually) all the details of internal machinery. And internals will change over time (for example, we used to not drop connection immediately on setAutoReconnect(false), making it less clear what events like that actually track / mean).

I'd rather say - clients can track this kind of timing only for very small set of scenarios, like boot, where they have full control over sequencing (when they call various APIs).

**Comment by vladsud** [comment]
> Other way to think about this - The host is interested in cases where the user is disconnected but doesn't want to be. Today, it's starting a 15 second timer after disconnected event. But that's not "fair" to us, because maybe we don't even try to connect until some time later. This event gives the host a clear picture - "I am trying to connect....I connected" will be the message they get now.

I'm not sure I follow. 15 second is UX thing, it has nothing to do with ability to track (internal) transitions.

> There is a case where host apps can take responsibility for calling resume and OWH doesn't know. This is what triggered this work. Hoping to get this in for next release (should have started it sooner) so it's ready for Vivek to code against when he returns in 2 weeks.

I think if OWH moves responsibility to call resume() to hosts, then responsibility to track all the times moves as well. Basically, user does not care if connection takes 5 seconds if hosts waits 10 seconds before calling resume(). You want to see first full picture (end-to-end), then follow with break down.

I do not have very big push back on event, but let's clearly define the meaning of it, as right now it's not clear to me when it should fire (see comment on IF check)

**Comment by markfields** [comment]
Closing for now, the partner asking for this event can build on `disconnected` event as the signal that we are (about to be) trying to connect.

**Pull request closed without merging**